### PR TITLE
Add Notepad3 5.19.108.1602

### PR DESCRIPTION
**Pull request by linsui**
Close https://github.com/lukesampson/scoop-extras/issues/2384

This is a portable version.

### CODE REVIEW

**Comment by linsui** [comment]
Maybe we can move Notepad2 and Notepad2-mod into Version?

File: bucket/notepad3.json [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        "Notepad3.exe",
+        "minipath.exe"
+    ],
+    "shortcuts": [
+        [
+            "Notepad3.exe",
+            "Notepad3"
+        ],
+        [
+            "minipath.exe",
```
Comment (Ash258) [inline bucket/notepad3.json]: What is this shortcut for?
This sounds like helper command rather than full gui app

File: bucket/notepad3.json [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    ],
+    "persist": [
+        "Favorites",
+        "minipath.ini",
+        "Notepad3.ini"
+    ],
+    "architecture": {
+        "64bit": {
+            "installer": {
+                "script": "Expand-7zipArchive \"$dir\\Notepad3_${version}_x64.zip\" \"$dir\""
```
Comment (Ash258) [inline bucket/notepad3.json]:
```suggestion
                "script": "Expand-7zipArchive \"$dir\\Notepad3_${version}_x64.zip\""
```

File: bucket/notepad3.json [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            }
+        },
+        "32bit": {
+            "installer": {
+                "script": "Expand-7zipArchive \"$dir\\Notepad3_${version}_x86.zip\" \"$dir\""
```
Comment (Ash258) [inline bucket/notepad3.json]:
```suggestion
                "script": "Expand-7zipArchive \"$dir\\Notepad3_${version}_x86.zip\""
```

File: bucket/notepad3.json (hunk 1 above)
Comment (linsui) [reply bucket/notepad3.json]: I don't know what it is, but it has a gui

**Comment by linsui** [comment]
Since Notepad2-mod has been archived, I remove auto update.

**Comment by linsui** [comment]
@Ash258 Please review.
I wonder what if there is space in `$dir`

```json
"    $notepad3 = \"$dir\\Notepad3.exe\".Replace('\\', '\\\\')",
```

File: bucket/notepad3.json [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            }
+        }
+    },
+    "post_install": [
+        "Remove-Item \"$dir\\Notepad3_${version}_x*.zip\"",
+        "if (Test-Path(\"$dir\\replace-windows-notepad.reg\")) {",
+        "    $notepad3 = \"$dir\\Notepad3.exe\".Replace('\\', '\\\\')",
+        "    $content = Get-Content \"$dir\\replace-windows-notepad.reg\"",
+        "    $content = $content.Replace('$notepad3', $notepad3)",
+        "    $content | Set-Content -Path \"$dir\\replace-windows-notepad.reg\"",
```
Comment (Ash258) [inline bucket/notepad3.json]: - No encoding
- Do not use pipe for this easy expressions

```suggestion
        "    Set-Content \"$dir\\replace-windows-notepad.reg\" $content -Encoding ASCII",
```

File: bucket/notepad3.json [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+                "script": "Expand-7zipArchive \"$dir\\Notepad3_${version}_x86.zip\""
+            }
+        }
+    },
+    "post_install": [
+        "Remove-Item \"$dir\\Notepad3_${version}_x*.zip\"",
+        "if (Test-Path(\"$dir\\replace-windows-notepad.reg\")) {",
+        "    $notepad3 = \"$dir\\Notepad3.exe\".Replace('\\', '\\\\')",
+        "    $content = Get-Content \"$dir\\replace-windows-notepad.reg\"",
+        "    $content = $content.Replace('$notepad3', $notepad3)",
```
Comment (Ash258) [inline bucket/notepad3.json]:
```suggestion
        "    $content = $content -replace '$notepad3', $notepad3",
```

File: bucket/notepad3.json [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        "32bit": {
+            "installer": {
+                "script": "Expand-7zipArchive \"$dir\\Notepad3_${version}_x86.zip\""
+            }
+        }
+    },
+    "post_install": [
+        "Remove-Item \"$dir\\Notepad3_${version}_x*.zip\"",
+        "if (Test-Path(\"$dir\\replace-windows-notepad.reg\")) {",
+        "    $notepad3 = \"$dir\\Notepad3.exe\".Replace('\\', '\\\\')",
```
Comment (Ash258) [inline bucket/notepad3.json]: Use powershell syntax

```suggestion
        "    $notepad3 = \"$dir\\Notepad3.exe\" -replace '\\', '\\\\'",
```

**Comment by Luckz** [comment]
FWIW there's a newer version, https://github.com/rizonesoft/Notepad3/releases/tag/RELEASE_5.19.630.2381

**Comment by linsui** [comment]
Sorry, I forgot to update hash...

**Comment by Ash258** [comment]
Why aren't you running checkver or let excavator do it?

**Comment by linsui** [comment]
And there is another bug. It seems that `-replace` is different from `.Replace()` and accept a regex as the second arg. So `\\\\` doesn't work.

**Comment by linsui** [comment]
In first arg of `-replace`, `\` need to be escape because it's a regex. But first arg of `.Replace()` is a string.

**Comment by Luckz** [comment]
> This is a portable version.

Would it make sense to split this into two, a `notepad3.json` that indeed just adds a "portable" application, and a `notepad3-replace.json` that also runs the .reg files (which will require admin permissions)?
Scoop does not offer any way of making this optional with a single JSON, right?
Another option would be downloading but not executing the .reg files. That's less automation, but the lost automation requires administrator permissions anyway.

Also, the path set for the .reg should perhaps really not be the "actual" scoop app folder but the exe inside the \current\ symlink folder which always points at the newest executable. Otherwise the .reg would have to be set again on every update, which again makes little sense on paper.

**Comment by linsui** [comment]
The reg file won't be run unless you run it manully. The path is set to current now.

**Pull request merged**